proposals/p1885.md

```diff
@@ -712,29 +712,23 @@ invalid value, to satisfy iterator comparison when `Next()` returns an empty
 `Optional`.
 
 Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
-need to cache the last value, to allow for `*it` to be called, and to be called
-more than once.
+need to copy and cache the last value, to allow for `*it` to be called. This
+puts a requirement on `ElementType` to be copyable for this interoperability to
+be usable.
```

Comment:
I think this will still work if `ElementType` is move-only (but see the comment below), it just means that `IterateIterator` will be move-only in those cases. That means it won't be a valid iterator according to the standard library, but it should still work with range-based `for` loops.

Reply:
I just looked back at the code: 
```
     if (const auto v = container_->Next(*cursor_); v) {
      value_ = *std::move(v);
    }
```
Won't `Next()` require `ElementType` to be copyable nonetheless, as we are effectively returning a copy of an element, and not moving anything from the container? Even if we avoid an optional-like wrapper, or just move from it, the `ElementType` still has to be copied at some point (likely withing `Next()`).

Maybe I forgot something between now and then.

Reply:
In recent versions of C++, if `Next` returns an `ElementType` rather than a reference, the code you quote won't actually copy or move the return value; it will use `v` itself as the storage for the returned temporary. Now, you're right that in this context, the _body_ of `Next` will very likely need to make a copy, but `Next` is written in pure Carbon, so that problem isn't caused by caching in the interoperability layer, and in fact it's not caused by the interop layer at all -- the same problem exists in pure Carbon.

Reply:
Agree that this is by design, and not caused by the interop. To expand, I think we could say "move-only" in the C++ domain, most that would be ignoring a very likely "copiable" requirement on Carbon side / `Next`. So all-in-all, "copyable" seem more correct, and the upper bound of reqs, vs the lower bound.

Reply:
The problem is that this doc doesn't explicitly talk about copyability on the Carbon side, so by introducing it here as a new topic that's specifically tied to caching in the interop layer, it really makes it sound like native Carbon will not require copyability, but C++ interop will.

I'd suggest revising the "Large elements and Optional" section to explicitly talk about the fact that this design may force `Next` to make a copy, and hence require `ElementType` to be copyable. Then down here you can say that the caching requires `ElementType` to be movable, and note that this won't matter in practice if Carbon requires it to be copyable anyway.